[E2E][JF] Add skeleton for steps 6-7 of JCM

doru91 opened the pull request:
Role of this PR is to introduce the main building blocks for executing steps 6-7 of JCM (see spec, Figure 93 - Joint Commissioning Steps and the associated text).

High-level architecture is described in [Slide 4](https://docs.google.com/presentation/d/1eOGmYHz6L1jsVTAbpDpZjaD-avM_KJ9L/edit#slide=id.p4) and the associated use-case (which will be fully implemented in following PRs) is in [Slides 5-7](https://docs.google.com/presentation/d/1eOGmYHz6L1jsVTAbpDpZjaD-avM_KJ9L/edit#slide=id.p5).

Associated SDK execution plan can be found in [project-chip/123](https://github.com/orgs/project-chip/projects/123/views/1), SDK section.

Fix https://github.com/project-chip/connectedhomeip/issues/38178, https://github.com/project-chip/connectedhomeip/issues/38201:
- PR is split in multiple commits with clear description for easier review
- first commit is doing some general refactoring around PAI/DAC certification generation/release to avoid code duplication. (that API will be used also for handling the JF certificates). Rest of the commissioner code required for steps 6-7 of JCM is guarded by CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC - by doing this we make sure we don't break any existing functionality
- during the read commissioning phase, the AdministratorFabricIndex, Fabric Table and Trusted Root Certificates are read, associated information from the Fabric Table are also read. Then, only the entry from the Fabric Table corresponding to AdministratorFabricIndex is saved. This info will be used in a following PR for Fabric Table VID Verification.
- more details about the architecture can be found in the JF guide - [docs/guides/joint_fabric_guide.md](https://github.com/doru91/connectedhomeip/blob/feature/jf/docs/guides/joint_fabric_guide.md)

### Testing
- used jf-control-app for pairing jf-admin-app as anchor:  `pairing onnetwork 1 110220033 --anchor true `
- used jf-control-app to OCW on jf-admin-app: `pairing open-commissioning-window 1 1 400 1000 1261` -> extract pairing_code
- used jf-control-app to test that the JF information is saved correctly: `pairing onnetwork 10 pairing_code --execute-jcm true`

doru91 committed:
autocommissioner: Refactor PAI/DAC certificate alloc/de-alloc

This avoids code duplication and the new certicate API may be used by JF.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] jf-control-app: Add --execute-jcm option

If this option is set during the pairing command, then JCM flow is triggered.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] core-sdk/jf-control-app: allow enabling/disabling JF support in core-SDK

Enable JF suppport by default when jf-control-app is being compiled.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 committed:
[E2E][JF] core-sdk/autocommissioner: add skeleton for executing steps 6-7 of JCM

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

restyled-commits committed:
Fix CI/CV issues/comments

* check that the commissioned Admin has an Admin CAT inside its NOC.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

tcarmelveilleux commented on src/platform/device.gni:
```diff
@@ -117,6 +117,11 @@ declare_args() {
 
+declare_args() {
+  # Enable Joint Fabric features
+  chip_device_config_enable_joint_fabric = false
```
There are no tests for the new code. Is that intended?

doru91 replied:
Indeed, no tests at this moment. @robertfarnum will follow-up immediately in a new PR with the tests and with some refactoring of the joint commissioning method state machine. Just to ease this ownership transfer, I would like if possible to signal the possible issues from the current code and allow us to introduce the tests in a follow up PR. Pls note that all the newly introduced code is under `CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC` and in order to introduce some tests there is a dependency on https://github.com/project-chip/connectedhomeip/pull/34637.

doru91 replied:
Resolving this conversation (also announced Tennessee on Slack).

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2777,6 +2794,187 @@ CHIP_ERROR DeviceCommissioner::ParseICDInfo(ReadCommissioningInfo & info)
+        if (!administratorFabricIndex.IsNull())
+        {
+            ChipLogProgress(Controller, "JCM: AdministratorFabricIndex: %d", administratorFabricIndex.Value());
+            info.JFAdministratorFabricIndex = administratorFabricIndex.Value();
+            info.JFAdminEndpointId          = path.mEndpointId;
```
Can there be more than one JointFabricAdministrator cluster on the node?  Or is there at most one, but we just don't know which endpoint it will be on?

doru91 replied:
Second case: there is a single JointFabricAdministrator cluster on the node but we don't know which endpoint it will be on.

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2777,6 +2794,187 @@ CHIP_ERROR DeviceCommissioner::ParseICDInfo(ReadCommissioningInfo & info)
+                        break;
+                    }
+                }
+                return CHIP_NO_ERROR;
```
This is never checking whether we got a value but it's not well-formed TLV (in which case Next() will return false at some point and the status of the iterator will be a failure).

doru91 replied:
Indeed, fixed this not by checking the iterator status but by making sure that the fields gets populated. In the end, if the fields don't get populated, it's not of interest what was the root issue of that.

bzbarsky-apple replied:
How do you want to handle the case when the fields get populated but the TLV is still invalid _after_ that point?

bzbarsky-apple replied:
I guess that works because you stop iterating at that point, ok.

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2777,6 +2794,187 @@ CHIP_ERROR DeviceCommissioner::ParseICDInfo(ReadCommissioningInfo & info)
+                        ChipLogProgress(Controller, "JCM: Successfully parsed the Administrator NOC and ICAC");
+                        break;
+                    }
+                }
```
Again, this is never checking for iteration failure.

doru91 replied:
Pls see my answer above.

bzbarsky-apple replied:
But in this case we are not in fact checking whether the fields got populated, are we?

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -2777,6 +2794,187 @@ CHIP_ERROR DeviceCommissioner::ParseICDInfo(ReadCommissioningInfo & info)
+                ReturnErrorOnFailure(this->mAttributeCache->Get<TrustedRootCertificates::TypeInfo>(path, trustedCAs));
+
+                auto iter = trustedCAs.begin();
+                while (iter.Next())
```
All these iterators need status checks at end of iteration.

doru91 replied:
Pls see my answer above.

bzbarsky-apple commented on src/controller/CommissioningDelegate.h:
```diff
@@ -629,6 +701,11 @@ class CommissioningParameters
         mICDSymmetricKey.ClearValue();
         mExtraReadPaths = decltype(mExtraReadPaths)();
+#if CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
+        mJFAdminNOC.ClearValue();
```
OK, but doesn't the AutoCommissioner code that makes copies of these external buffer things then need to do that?

bzbarsky-apple replied:
Ah, I guess these things are not actually used as input to AutoCommissioner in a meaningful way?

As in, if the API consumer sets these on its incoming commissioning parameters they will be ignored?

doru91 replied:
Actually my understanding is that these things _are_ actually used as input to AutoCommissioner in a meangful way.

 `ClearExternalBufferDependentValues` gets called by `AutoCommissioner::SetCommissioningParameters(const CommissioningParameters & params)` in the Autocommissioner constructor and the lines of code that I think answer your question are:

```
CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParameters & params)
{
    // Our logic below assumes that we can modify mParams without affecting params.
    VerifyOrReturnError(&params != &mParams, CHIP_NO_ERROR);

    // Copy the whole struct (scalars and pointers), but clear any members that might point to
    // external buffers. For those members we have to copy the data over into our own buffers below.
    // Note that all of the copy operations use memmove() instead of memcpy(), because the caller
    // may be passing a modified shallow copy of our CommissioningParmeters, i.e. where various spans
    // already point into the buffers we're copying into, and memcpy() with overlapping buffers is UB.
    mParams = params;
    mParams.ClearExternalBufferDependentValues();
```

So the Autocommissioner has its own CommissioningParameters instance (as the mParams field) which gets populated from the `params` input parameter. Before copying from params to mParams we want to make sure that mParams holds no dangling pointers.

bzbarsky-apple replied:
My point, which was not very clear perhaps, is: we are not passing these things into AutoCommissioner via the CommissioningParameters argument to SetCommissioningParameters.  If the input CommissioningParameters has a mJFAdminNOC, we will ignore it, because of the ClearExternalBufferDependentValues() call.

doru91 replied:
That's correct, ClearExternalBufferDependentValues() makes sure that if mJFAdminNOC from the input CommissioningParameters holds a value then the associated memory will be freed - so we are on the same page. mJFAdminNOC will be populated during CommissioningStage::kReadCommissioningInfo. 

As a side note, I don't fully understand the code from SetCommissioningParameters:

```
CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParameters & params)
{
    // Our logic below assumes that we can modify mParams without affecting params.
    VerifyOrReturnError(&params != &mParams, CHIP_NO_ERROR);

    // Copy the whole struct (scalars and pointers), but clear any members that might point to
    // external buffers. For those members we have to copy the data over into our own buffers below.
    // Note that all of the copy operations use memmove() instead of memcpy(), because the caller
    // may be passing a modified shallow copy of our CommissioningParmeters, i.e. where various spans
    // already point into the buffers we're copying into, and memcpy() with overlapping buffers is UB.
    mParams = params;
    mParams.ClearExternalBufferDependentValues();

    if (params.GetThreadOperationalDataset().HasValue())
    {
        ByteSpan dataset = params.GetThreadOperationalDataset().Value();
        if (dataset.size() > CommissioningParameters::kMaxThreadDatasetLen)
        {
            ChipLogError(Controller, "Thread operational data set is too large");
            return CHIP_ERROR_INVALID_ARGUMENT;
        }
        memmove(mThreadOperationalDataset, dataset.data(), dataset.size());
        ChipLogProgress(Controller, "Setting thread operational dataset from parameters");
        mParams.SetThreadOperationalDataset(ByteSpan(mThreadOperationalDataset, dataset.size()));
    }
```

In ClearExternalBufferDependentValues() the value of mWiFiCreds from the input CommissioningParameters gets cleared so `params.GetThreadOperationalDataset().HasValue()` above should always be false, right?

bzbarsky-apple replied:
Right, exactly.

doru91 committed:
[E2E][JF] Fix issues signaled during the review process

* fix the error path when iterating through the attribute cache;
* use a ScopedMemoryBUfferWithSize for DAC, PAI and JF certs.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 commented on src/controller/CommissioningDelegate.h:
```diff
@@ -679,6 +755,25 @@ class CommissioningParameters
+    Optional<uint16_t> mJFAdminICACLen;
+
+    Optional<ByteSpan> mJFAdminRCAC;
+    Optional<uint16_t> mJFAdminRCACLen;
```
`uint16_t` has been replaced by `size_t`

doru91 replied:
Resolving this conversation (also announced Tennessee on Slack).

doru91 commented on src/controller/CommissioningDelegate.h:
```diff
@@ -777,6 +872,15 @@ struct ICDManagementClusterInfo
+#if CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
+struct JFFabricTableInfo
+{
+    VendorId vendorId = VendorId::Common;
+    FabricId fabricId = kUndefinedFabricId;
```
Agree is not unique, but I am saving it for later usage for the Fabric Table Vendor ID Verification. fabricIndex is saved is the one corresponding to `JFAdministratorFabricIndex`.

doru91 replied:
Resolving this conversation (also announced Tennessee on Slack).

doru91 commented on src/controller/AutoCommissioner.h:
```diff
@@ -65,11 +65,9 @@ class AutoCommissioner : public CommissioningDelegate
-    CHIP_ERROR SetDAC(const ByteSpan & dac);
-    CHIP_ERROR SetPAI(const ByteSpan & pai);
+    CHIP_ERROR SaveCertificate(ByteSpan inCertSpan, uint8_t ** outCert, uint16_t * outCertSize);
+    void ReleaseCertificate(uint8_t ** cert, uint16_t * certSize);
```
Indeed, the old code was missing some ReleaseCertificate calls. 

Changed to using ScopedMemoryBufferWithSize for the DAC/PAI and for the JF certs. Hopefully I did it in a correct way, pls double-check.

doru91 committed:
[E2E][JF] Initialize AdministratorFabricIndex

Set AdministratorFabricIndex on jf-admin-app immediately after
its onboarding by jf-control-app.

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

vijs commented on examples/jf-control-app/commands/pairing/PairingCommand.h:
```diff
@@ -106,6 +106,7 @@ class PairingCommand : public CHIPCommand,
             break;
         case PairingMode::Code:
             AddArgument("anchor", 0, 1, &mAnchor);
+            AddArgument("execute-jcm", 0, 1, &mExecuteJCM);
```
Can we move the "execute-jcm" and "anchor" arguments into the PairingCommand class constructor, since they're common to all subcommands?

doru91 replied:
Thanks for this suggestion. Pushed the fix, pls check.

doru91 committed:
[E2E][JF] Refactor pairing command parameters location

Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

bzbarsky-apple commented on examples/jf-control-app/args.gni:
```diff
@@ -34,3 +34,7 @@ matter_log_json_payload_decode_full = true
 chip_tlv_validate_char_string_on_read = true
 chip_tlv_validate_char_string_on_write = true
+
+# enable Joint Fabric features in the code SDK code
```
```suggestion
# enable Joint Fabric features in the core SDK code
```

bzbarsky-apple commented on src/app/clusters/joint-fabric-administrator-server/joint-fabric-administrator-server.cpp:
```diff
@@ -93,7 +93,7 @@ CHIP_ERROR JointFabricAdministratorAttrAccess::Read(const ConcreteReadAttributeP
 CHIP_ERROR JointFabricAdministratorAttrAccess::ReadAdministratorFabricIndex(AttributeValueEncoder & aEncoder)
 {
-    return aEncoder.Encode(Server::GetInstance().GetJointFabricDatastore().GetAdministratorFabricIndex());
+    return CHIP_NO_ERROR;
```
Why does this function exist, then?

But also, it's weird that we are not getting this from the joint fabric data store.  Why aren't we?

doru91 replied:
JointFabricAdministratorAttrAccess::ReadAdministratorFabricIndex shouldn't exist, it was wrongly introduced by https://github.com/project-chip/connectedhomeip/pull/34637 and then https://github.com/project-chip/connectedhomeip/pull/38842 inexplicable saved the mAdministratorFabricIndex in the Joint Fabric Datastore -by doing this instead of reading the value of mAdministratorFabricIndex@JFAdministratorFabricIndex, its saved value from JF DS implementaton is returned. I opened https://github.com/project-chip/connectedhomeip/issues/39101 to fix this in a separate PR.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -777,17 +785,76 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
+                {
+                    ChipLogError(Controller, "AutoCommissioner: cannot allocate memory for mJFAdminRCAC");
+                    return CHIP_ERROR_NO_MEMORY;
+                    ;
```
Why?

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -777,17 +785,76 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
+                memcpy(mJFAdminRCAC.Get(), mDeviceCommissioningInfo.JFAdminRCAC.data(), mJFAdminRCAC.AllocatedSize());
+                mParams.SetJFAdminRCAC(ByteSpan(mJFAdminRCAC.Get(), mJFAdminRCAC.AllocatedSize()));
+
+                if (!mJFAdminICAC.Alloc(mDeviceCommissioningInfo.JFAdminICAC.size()))
```
I assume that size is known to be nonzero here?

doru91 replied:
mDeviceCommissioningInfo.JFAdminICAC.size() represents the size of the peer Admin's ICAC, as read in ParseJFAdministratorFabricInfo (_info.JFAdminICAC = nocStruct.icac.Value()_). I think the size could be 0 if the peer Admin is a rogue device, but the ICAC validity will be checked later in the process (as a side note, this is pretty similar to the flow that already exists for reportPAISpan from Autocommissioner.cpp).

doru91 replied:
Added the check for 0 size.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -777,17 +785,76 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
+                {
+                    ChipLogError(Controller, "AutoCommissioner: cannot allocate memory for mJFAdminICAC");
+                    return CHIP_ERROR_NO_MEMORY;
+                    ;
```
?

doru91 replied:
removed it.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -777,17 +785,76 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
+                {
+                    ChipLogError(Controller, "AutoCommissioner: cannot allocate memory for mJFAdminNOC");
+                    return CHIP_ERROR_NO_MEMORY;
+                    ;
```
?

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -777,17 +785,76 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
+        case CommissioningStage::kSendPAICertificateRequest: {
+            auto reportPAISpan = report.Get<RequestedCertificate>().certificate;
+
+            if (!mPAI.Alloc(reportPAISpan.size()))
+            {
+                ChipLogError(Controller, "AutoCommissioner cannot allocate memory for mPAI");
+                return CHIP_ERROR_NO_MEMORY;
+            }
+
+            memcpy(mPAI.Get(), reportPAISpan.data(), mPAI.AllocatedSize());
```
We have a bunch of copies of this code.  Can we put it into a helper function, please?

doru91 replied:
Yes, I created AllocateMemoryAndCopySpan() helper function.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -777,17 +785,76 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
+                    ;
+                }
+                memcpy(mJFAdminRCAC.Get(), mDeviceCommissioningInfo.JFAdminRCAC.data(), mJFAdminRCAC.AllocatedSize());
+                mParams.SetJFAdminRCAC(ByteSpan(mJFAdminRCAC.Get(), mJFAdminRCAC.AllocatedSize()));
```
```suggestion
                mParams.SetJFAdminRCAC(mJFAdminRCAC.Span());
```

and similar elsewhere?

doru91 replied:
Applied all these suggestions, resolving this conversation.

bzbarsky-apple commented on src/controller/CommissioningDelegate.h:
```diff
@@ -51,6 +51,10 @@ enum CommissioningStage : uint8_t
     kSendAttestationRequest,     ///< Send AttestationRequest (0x3E:0) command to the device
     kAttestationVerification,    ///< Verify AttestationResponse (0x3E:1) validity
     kAttestationRevocationCheck, ///< Verify Revocation Status of device's DAC chain
+#if CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
+    kJFValidateNOC,              ///< Verify Admin NOC contains an Administrator CAT
+    kSendVIDVerificationRequest, ///< Send SignVIDVerificationRequest command to the device
+#endif                           // CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
```
We should probably put these at the end of the enum, so the numbering of the other bits does not depend on whether JF is enabled?

doru91 replied:
Enum follows the transitions of the commissioner state machine so it will be a little bit confusing if I put those JF bits at the end - I can do it if you want to. Or maybe I can keep those value present even if CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC is not defined?

bzbarsky-apple replied:
Let's just keep the values present no matter what.  There is no cost to it, and it simplifies things a lot.

bzbarsky-apple replied:
Can we leave the cases in the switch even when JF is not defined, but just condition the actual logic on JF?

It's really important for client sanity to have stable values for this enumeration for all the other steps that don't depend on whether JF is enabled.

doru91 committed:
[E2E][JF] Fix review comments

* main change is related to creation of AllocateMemoryAndCopySpan
* small cosmetic fixes

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>
Signed-off-by: Doru Gucea <doru-cristian.gucea@nxp.com>

doru91 commented on src/controller/AutoCommissioner.h:
```diff
@@ -65,14 +65,9 @@ class AutoCommissioner : public CommissioningDelegate
-
-    ByteSpan GetDAC() const { return ByteSpan(mDAC, mDACLen); }
-    ByteSpan GetPAI() const { return ByteSpan(mPAI, mPAILen); }
+    ByteSpan GetDAC() const { return ByteSpan(mDAC.Get(), mDAC.AllocatedSize()); }
```
Applied this suggestion.

bzbarsky-apple commented on src/controller/AutoCommissioner.h:
```diff
@@ -113,6 +108,17 @@ class AutoCommissioner : public CommissioningDelegate
     void ResetTryingSecondaryNetwork() { mTryingSecondaryNetwork = false; }
     bool mTryingSecondaryNetwork = false;
 
+    CHIP_ERROR AllocateMemoryAndCopySpan(Platform::ScopedMemoryBufferWithSize<uint8_t> & scopedBuffer, ByteSpan span)
```
Why is this inlined?  I'd think we want this in the .cpp file.

doru91 replied:
moved it in the .cpp file

bzbarsky-apple commented on src/controller/AutoCommissioner.h:
```diff
@@ -113,6 +108,17 @@ class AutoCommissioner : public CommissioningDelegate
 
+    CHIP_ERROR AllocateMemoryAndCopySpan(Platform::ScopedMemoryBufferWithSize<uint8_t> & scopedBuffer, ByteSpan span)
+    {
+        if (!scopedBuffer.Alloc(span.size()))
```
My point in https://github.com/project-chip/connectedhomeip/pull/38749#discussion_r2096091204 is that allocating a 0-sized buffer will crash with some libc implementations.

So ideally we should check for `span.size() == 0` and if so just `scopedBuffer.Free()` and return.

doru91 replied:
Added the check for 0 size.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -927,77 +967,19 @@ CHIP_ERROR AutoCommissioner::PerformStep(CommissioningStage nextStage)
+    if (!span.size())
     {
-        chip::Platform::MemoryFree(mPAI);
+        return CHIP_ERROR_INVALID_MESSAGE_LENGTH;
```
So for JF there is always required to be an ICAC, I guess?

doru91 replied:
That's right, an ICAC is mandatory for JF.

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -3354,6 +3560,40 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
     break;
+
+    case CommissioningStage::kJFValidateNOC:
+#if CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
```
This is doing implicit fallthroughs to kSendOpCertSigningRequest when CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC is not defined.  Shouldn't it `CommissioningStageComplete` with some error instead?

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -3354,6 +3560,40 @@ void DeviceCommissioner::PerformCommissioningStep(DeviceProxy * proxy, Commissio
+#endif
+
+    case CommissioningStage::kSendVIDVerificationRequest:
+#if CHIP_DEVICE_CONFIG_ENABLE_JOINT_FABRIC
```
Again, we should not be doing fall-through here.
